### PR TITLE
Highlower

### PR DESCRIPTION
**Pull request by erxclau**

### CODE REVIEW

File: app/app.py [hunk 1]
```diff
@@ -110,9 +113,57 @@ def callback():
...
+    return render_template(
+        'higherlowergame.html',
+        songs=songs
+        )
+
+@app.route("/save_song/<song_id>")
+def save_song(song_id):
+    if not 'access_token' in session:
+        flash('You are not connected to your Spotify account', 'error')
+        return redirect(url_for('home'))
```
Comment (raymondlee00) [inline app/app.py]: @login_required decorator

Comment (erxclau) [reply app/app.py]: the access token check acts as a login decorator

File: app/app.py [hunk 2]
```diff
@@ -96,6 +97,15 @@ def callback():
 
     session['access_token'] = access_token
 
+    get_user_info()
+
+    return redirect(url_for('home'))
+
+def get_user_info():
+    get_user_name()
+    get_user_top()
```
Comment (erxclau) [inline app/app.py]: nix

File: app/app.py [hunk 3]
```diff
@@ -110,8 +120,66 @@ def callback():
...
+    data = json.loads(res)['items']
+
+    songs = list()
+    for track in data:
+        track_data = {
+            'title': track['name'],
+            'artist': track['album']['artists'][0]['name'],
+            'coverArtLink': track['album']['images'][0]['url'],
+            'genre': "I DONT KNOW",
+            'lyrics': "MUSIXMATCH",
```
Comment (erxclau) [inline app/app.py:155]: needs fixing

File: app/app.py [hunk 4]
```diff
@@ -110,8 +120,66 @@ def callback():
...
+            'popularity': track['popularity'],
+            'spotify_id': track['id'],
+            'iframe': f"{track['external_urls']['spotify'][:25]}embed/{track['external_urls']['spotify'][25:]}",
+        }
+        songs.append(track_data)
+    session['songs'] = songs
```
Comment (erxclau) [inline app/app.py]: convert to db operation

File: app/app.py [hunk 5]
```diff
@@ -110,8 +120,66 @@ def callback():
...
+        req = urllib.request.Request(
+            f"https://api.spotify.com/v1/me/tracks?ids={song_id}",
+            headers=authorization_header,
+            method="PUT",
+        )
+
+        try:
+            req = urllib.request.urlopen(req)
+        except urllib.error.HTTPError as error:
+            print(error.reason)
```
Comment (erxclau) [inline app/app.py]: i dont think this try exception block is necessary anymore

File: app/app.py [hunk 6]
```diff
@@ -127,17 +195,27 @@ def hearted_songs():
...
             "hearted_songs.html",
             data = data['items'],
-    )
+        )
+
+@app.route('/logout')
+def logout():
+    session['access_token'] = None
+    session['display_name'] = None
+    session['spotify_client_id'] = None
```
Comment (erxclau) [inline app/app.py]: why do we store client id

File: app/templates/home.html [hunk 7]
```diff
@@ -1,99 +1,138 @@
-{% extends "base.html" %}
-
-{% block title %}
-Home
-{% endblock %}
-
-{% block background %}
+{% extends "base.html" %} {% block title %} Home {% endblock %} {% block
+background %}
```
Comment (erxclau) [inline app/templates/home.html]: ugly

File: app/templates/home.html [hunk 8]
```diff
@@ -1,99 +1,138 @@
...
+            <img class="bottom-image pt-2" src="../static/img/heart.gif" />
+            <br />
+            <p class="h5 pt-2">My Hearted Songs</p>
+            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
+            eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
+            ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
+            aliquip ex ea commodo consequat. Duis aute irure dolor in
+            reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
+            pariatur. Excepteur sint occaecat cupidatat non proident, sunt in
+            culpa qui officia deserunt mollit anim id est laborum.
```
Comment (erxclau) [inline app/templates/home.html]: replace lorem ipsum

File: app/app.py (hunk 1 above)
Comment (raymondlee00) [reply app/app.py]: okie

File: app/app.py (hunk 6 above)
Comment (raymondlee00) [reply app/app.py]: dont

**Review by erxclau — COMMENTED** [review]
spotify api queries can be abstracted away NTTSTT with auth headers, urlopen, etc

File: app/app.py (hunk 3 above)
Comment (yaruluo) [reply app/app.py:155]: h s

**Review by yaruluo — COMMENTED** [review]
wonderful

**Review by MattChann — APPROVED** [review]

**Pull request merged**